docs/proposals-accepted/202209-receive-tenant-external-labels.md

```diff
@@ -0,0 +1,72 @@
+---
+type: proposal
+title: Allow specifying tenant-specific external labels in Receivers
+status: accepted
+owner: haanhvu
+menu: proposals-accepted
+---
+
+## 1 Related links/tickets
+
+https://github.com/thanos-io/thanos/issues/5434
+
+## 2 Why
+
+We would like to select and query tenant blocks from Receivers in an arbitrary manner.
+
+## 3 Pitfalls of the current solution
+
+Currently, we can only add external labels to Receiver itself, not to each tenant in the Receiver. There’s currently no way to select and query tenant blocks in an arbitrary manner.
+
+## 4 Goals
+
+* Allow users to add arbitrary tenants’ external labels in the easiest way possible
+* Allow users to change arbitrary tenants’ external labels in the easiest way possible
+* Changes in tenants’ external labels are handled correctly
+* Backward compatibility (e.g., with Receiver’s external labels) is assured
+* Tenants’ external labels are handled separately in RouterIngestor, RouterOnly, and IngestorOnly modes
+
+## 5 Non-goals
+
+* Logically split RouterOnly and IngestorOnly modes (Issue [#5643](https://github.com/thanos-io/thanos/issues/5643)): If RouterOnly and IngestorOnly modes are logically split, implementing tenants’ external labels in RouterOnly and IngestorOnly modes would be less challenging. However, the split could be tricky and time-consuming, which might distract us from implementing tenants’ external labels. So we will not do it now.
+
+## 6 Audience
+
+Users who need to have cross-tenant information and do cross-tenant activities
+
+## 7 How
+
+We’ll add a new `external_labels` field in the hasring config. Something like this:
+ ```
+ [
+    {
+        "hashring": "tenant-a",
+        "endpoints": ["tenant-a-1.metrics.local:19291/api/v1/receive", "tenant-a-2.metrics.local:19291/api/v1/receive"],
+        "tenants": ["tenant-a"]
+        "external_labels": ["label-1", "label-2", "label-3"]
```

Comment:
@fpetkovski @saswatamcode here should it be `"external_labels": ["label-name-1:label-value-1", "label-name-2:label-value-2", "label-name-3:label-value-3"]`? 🤔

Reply:
I think it can be similar to how we pass in label key-value pairs via flag like `--label='receive_replica="0"'`.
So it would be something like `external_labels": ["label=value", "label2=value2"]`.

Reply:
Yes, so I think you can reuse `parseFlagLabels` https://github.com/thanos-io/thanos/blob/main/cmd/thanos/rule.go#L747 to basically parse string slice from YAML.

Reply:
@saswatamcode Is the quote surrounding the label value important?

Reply:
So `"external_labels": ["label-1", "label-2", "label-3"]` is specifying the label name value pairs? or just label names?